adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+------------------------ MODULE query_deduplication ------------------------
+
+
+EXTENDS Integers, Sequences, FiniteSets, TLC
+
+CONSTANT Proc
+CONSTANT Keys
+(*
+
+\* from https://probablydance.com/2020/10/31/using-tla-in-the-real-world-to-understand-a-glibc-bug/
+--algorithm Spinlock 
+{
+  variables
+    cache_lock = 0;
+    wait_map_lock = 1;
+    locks = [i \in {cache_lock, wait_map_lock} |-> FALSE];
+    \* set to false for every possible key at first
+    cache = [ i \in Keys |-> FALSE ];
+    wait_map = [ i \in Keys |-> {} ];
+    channels = [i \in Proc |-> FALSE ];
+    tested_keys = {};
+ 
+   define {
+
+    MutualExclusion == \A i, j \in Proc : 
+                         (i # j) => ~ /\ pc[i] = "cs"
+                                      /\ pc[j] = "cs"
+                                      
+    DeadlockFreedom == 
+        \A i \in Proc : 
+          (pc[i] = "exchange") ~> (\E j \in Proc : pc[j] = "cs")
+    
+    
+    \*AllFinished == <>(\A proc \in Proc: pc[proc] = "Done")
+    \*AllKeysLoaded == <>(\A k \in tested_keys: cache[k])
+  };
+
+  procedure lock(idx = 0)
+  {
+    exchange:
+      await ~locks[idx];
+    check:
+      if(~locks[idx]) {
+        locks[idx] := TRUE;
+        return;
+      } else {
+        goto exchange;
+      };
+  };
+
+  procedure unlock(idx = 0)
+  {
+    reset_state:
+      locks[idx] := FALSE;
+      return;
+  }
+  
+  fair process (P \in Proc)
+  variables Key \in Keys, pr = 0;
+  {
+  
+ 
+    start:
+        \* let mut locked_cache = self.cached.lock().await;
+        call lock(cache_lock);
```

Comment:
we begin by locking the cache. I put the corresponding code in comments, extracted from:
https://github.com/apollographql/router/blob/8484fd0662aab47c690e53e8d0f050ee157ac324/apollo-router-core/src/cache.rs#L50-L119